File: 121/121.dart
// Runtime
// 416
// ms
// Beats
// 34.17%
// of users with Dart
// Memory
// 192.17
// MB
// Beats
// 26.67%
// of users with Dart

class Solution {
  int maxProfit(List<int> prices) {
    if (prices.isEmpty) return 0;

    int maxProfit = 0;
    int minPrice = prices[0];

    for (int i = 1; i < prices.length; i++) {
      maxProfit =
          prices[i] - minPrice > maxProfit ? prices[i] - minPrice : maxProfit;
      minPrice = prices[i] < minPrice ? prices[i] : minPrice;
    }
    return maxProfit;
  }
}

main() {
  Solution sol = new Solution();
  print(sol.maxProfit([7, 1, 5, 3, 6, 4]));
  print(sol.maxProfit([7, 6, 4, 3, 1]));
  // wild test case
  print(sol.maxProfit([1, 10000, 1, 10000, 1, 10000, 1, 10000, 1, 10000]));
}

File: 121/ts/121.ts
// Runtime
// 83
// ms
// Beats
// 39.06 %
//     of users with TypeScript
// Memory
// 59.70
// MB
// Beats
// 54.40 %
//     of users with TypeScript

function maxProfit_sol(prices: number[]): number {
    var maxProfit = 0;
    var minPrice = prices[0];
    for (let i = 1; i < prices.length; i++) {
        minPrice = Math.min(minPrice, prices[i]);
        maxProfit = Math.max(maxProfit, prices[i] - minPrice);
    }
    return maxProfit;
};